Sources/NIOOpenSSL/ByteBufferBIO.swift

```diff
@@ -0,0 +1,302 @@
+    /// (that is, that consume a reference count).
+    ///
+    /// Note that the BIO may not remain useful for long periods of time: if the `ByteBufferBIO`
+    /// object that owns the BIO goes out of scope, the BIO will have its pointers invalidated
+    /// and will no longer be able to send/receive data.
+    internal func retainedBIO() -> UnsafeMutablePointer<BIO> {
+        CNIOOpenSSL_BIO_up_ref(self.bioPtr)
+        return self.bioPtr
+    }
+
+
+    /// Called to obtain the outbound ciphertext written by OpenSSL.
+    ///
+    /// This function obtains a buffer of ciphertext that needs to be written to the network. In a
+    /// normal application, this should be obtained on a call to `flush`. If no bytes have been flushed
+    /// to the network, then this call will return `nil` rather than an empty byte buffer, to help signal
+    /// that the `write` call should be elided.
+    ///
+    /// - returns: A buffer of ciphertext to send to the network, or `nil` if no buffer is available.
+    func outboundCiphertext() -> ByteBuffer? {
+        guard self.outboundBuffer.readableBytes > 0 else {
+            // No data to send.
+            return nil
+        }
+
+        /// Once we return from this function, we want to account for the bytes we've handed off.
+        defer {
+            self.outboundBuffer.moveReaderIndex(to: self.outboundBuffer.writerIndex)
+        }
+
+        return self.outboundBuffer
+    }
+
+    /// Stores a buffer received from the network for delivery to OpenSSL.
+    ///
+    /// Whenever a buffer is received from the network, it is passed to the BIO via this function
+    /// call. Only one buffer may be passed to OpenSSL at any one time: once a buffer is passed, it
+    /// is expected to be cleared immediately. If it is not cleared, this is an application error and
+    /// must be resolved.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer of ciphertext bytes received from the network.
+    func receiveFromNetwork(buffer: ByteBuffer) {
+        precondition(self.inboundBuffer == nil, "Did not flush inbound bytes to OpenSSL")
+        self.inboundBuffer = buffer
+    }
+
+    /// OpenSSL has requested to read ciphertext bytes from the network.
+    ///
+    /// This function is invoked whenever OpenSSL is looking to read data.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer for NIO to copy bytes into.
+    /// - returns: The number of bytes we have copied.
+    fileprivate func sslRead(buffer: UnsafeMutableRawBufferPointer) -> CInt {
+        guard var inboundBuffer = self.inboundBuffer else {
+            // We have no bytes to read. Mark this as "needs read retry".
+            CNIOOpenSSL_BIO_set_retry_read(self.bioPtr)
+            return -1
+        }
+
+        let bytesToCopy = min(buffer.count, inboundBuffer.readableBytes)
+        _ = inboundBuffer.readWithUnsafeReadableBytes { bytePointer in
+            assert(bytePointer.count >= bytesToCopy)
```

Comment:
feel free to ignore, in these conditions it's usually good to see the values should this ever actually trigger so maybe a string which shows the values?